### ISSUE TITLE
Provide a Sphinx directive to insert a mini-gallery

### ISSUE THREAD

**Issue by ayshih** [sphinx-gallery/sphinx-gallery#683]
In SunPy, we've started using [mini-galleries](https://sphinx-gallery.github.io/stable/configuration.html#add-mini-galleries-for-api-documentation), but I found the procedure for insertion to be a little clunky, and the mini-gallery to be a little limiting.  I wrote [a Sphinx directive for SunPy](https://github.com/ayshih/sunpy/blob/c7f8cb2776acb21a9029e943c868503772740b2b/docs/conf.py#L320), with the following advantages:

- No need to specify the backreferences directory or the `.examples` suffix
- No potential omission of the `<div class="sphx-glr-clear">` HTML block
- Removes the heading from the backreferences file (in case it doesn't match the desired heading level)
- If multiple inputs are provided (separated by spaces), the output mini-gallery concatenates all of the relevant examples

Live example ([found here](https://docs.sunpy.org/en/latest/code_ref/coordinates/rotatedsunframe.html#example-uses-of-rotatedsunframe)):
```rst
.. minigallery:: sunpy.coordinates.RotatedSunFrame
```

Combining multiple (possibly unrelated) objects into single mini-gallery:
```rst
.. minigallery:: sunpy.coordinates.sun.earth_distance sunpy.coordinates.transform_with_sun_center
```

Useful for when the same object can be accessed from multiple namespaces:
```rst
.. minigallery:: sunpy.coordinates.RotatedSunFrame sunpy.coordinates.metaframes.RotatedSunFrame
```

Would `sphinx-gallery` be interested in including such a directive (i.e, upstreamed from SunPy)?  It would need to be modified to be a bit more general and robust, of course.  Looking at some open issues, it looks like there may be interest (#38, #413), but also possibly a rule to not provide directives in this project?

**Comment by choldgraf** [comment]
I'm definitely a fan of exposing some of the more useful SG features as user-facing features such as this minigallery example. 👍

**Comment by larsoner** [comment]
> Would sphinx-gallery be interested in including such a directive (i.e, upstreamed from SunPy)? It would need to be modified to be a bit more general and robust, of course. Looking at some open issues, it looks like there may be interest (#38, #413), but also possibly a rule to not provide directives in this project?

No rule or anything prohibiting directives, just no compelling use cases thus far. This seems like it could be a good one!

Maybe start with a tiny PR just to `doc/configuration.rst` with what you would write there about the feature. That will force us to sort out the user-facing API and also figure out how it integrates with / instead of / in addition to the currently recommended way of doing things here:

https://sphinx-gallery.github.io/stable/configuration.html#auto-documenting-your-api-with-links-to-examples

For example, do you propose that we should replace all occurences of:
```
   .. include:: backreferences/{{fullname}}.{{item}}.examples

   .. raw:: html

	       <div class="sphx-glr-clear"></div>
```
With:
```
    .. minigallery:: {{fullname}}.{{item}}
```
? The old way would still work, but this is much more user-friendly so sounds good to me!

Once we sort that bit out, then you could actually implement with changes to our internal `doc/_templates`, and also `sphinx_gallery/tests/tinybuild/` and `sphinx_gallery/tests/test_full.py` to test it. WDYT?

**Comment by ayshih** [comment]
> No rule or anything prohibiting directives, just no compelling use cases thus far.

Great!  I got that impression from comments like [this](https://github.com/sphinx-gallery/sphinx-gallery/issues/413#issuecomment-417691567), which seemed to hint at some historical debate over providing directives.

> Maybe start with a tiny PR just to `doc/configuration.rst` with what you would write there about the feature.

Sure, I can do that (later this week).

For your example, as currently implemented in SunPy:
```rst
.. minigallery:: {{fullname}}.{{item}}
```
would be equivalent to:
```rst
.. include:: backreferences/{{fullname}}.{{item}}.examples
    :start-line: 5

.. raw:: html
    <div class="sphx-glr-clear"></div>
```

The `:start-line: 5` is needed to strip the hard-coded heading in the generated backreferences file.  The heading can conflict with the existing heading hierarchy where you want to embed the mini-gallery, and stripping the heading allows for seamless concatenation of multiple backreferences files into a single mini-gallery.  It's a bit of hack that could certainly be done differently in `sphinx-gallery` proper – one could imagine having a configuration option to turn off the headings – but that could break compatibility in ways you'd want to avoid.

**Comment by larsoner** [comment]
> Great! I got that impression from comments like this, which seemed to hint at some historical debate over providing directives.

Sphinx-gallery has been (trying to, at least) be more modular in the last couple of years, and nobody has complained so far. So I'd say you should be good to go on this one

> The :start-line: 5 is needed to strip the hard-coded heading in the generated backreferences file. The heading can conflict with the existing heading hierarchy where you want to embed the mini-gallery,

Indeed it sounds like this is a complaint from #413, and your directive seems like it would solve this problem

> one could imagine having a configuration option to turn off the headings – but that could break compatibility in ways you'd want to avoid (e.g., having the "old way" still work the same).

I think it's indeed easier just to know what we do and compensate appropriately. There is a cost to adding more and more configuration settings, so if we can avoid it and keep track of where we need to jump to internally we should. And keeping backward compatibility/behavior is an important goal.

Maybe just make sure in a code comment (above wherever the `Examples` heading is generated) that any changes to the template string (e.g., adding or removing newlines) can require refactoring of the `.. minigallery` directive to compensate for where to start the content.

**Comment by lucyleeow** [comment]
closed via #685
